Content.Server/Advertise/AdvertiseSystem.cs

```diff
@@ -32,102 +35,111 @@ public override void Initialize()
+        public void SetEnabled(EntityUid uid, bool enable, AdvertiseComponent? advertise = null)
         {
             if (!Resolve(uid, ref advertise))
                 return;
 
-            var attemptEvent = new AdvertiseEnableChangeAttemptEvent(enabled, advertise.Enabled);
-            RaiseLocalEvent(uid, attemptEvent, false);
+            if (advertise.Enabled == enable)
+                return;
+
+            var attemptEvent = new AdvertiseEnableChangeAttemptEvent(enable);
+            RaiseLocalEvent(uid, attemptEvent);
 
             if (attemptEvent.Cancelled)
                 return;
 
-            if(enabled)
-                RefreshTimer(uid, !advertise.Enabled, advertise);
-
-            advertise.Enabled = enabled;
+            advertise.Enabled = enable;
+            RefreshTimer(uid, advertise);
         }
 
-        private void OnPowerReceiverEnableChangeAttempt(EntityUid uid, ApcPowerReceiverComponent component, AdvertiseEnableChangeAttemptEvent args)
+        private static void OnPowerReceiverEnableChangeAttempt(EntityUid uid, ApcPowerReceiverComponent component, AdvertiseEnableChangeAttemptEvent args)
         {
-            if(args.NewState && !component.Powered)
+            if(args.Enabling && !component.Powered)
                 args.Cancel();
         }
 
-        private void OnVendingEnableChangeAttempt(EntityUid uid, VendingMachineComponent component, AdvertiseEnableChangeAttemptEvent args)
+        private static void OnVendingEnableChangeAttempt(EntityUid uid, VendingMachineComponent component, AdvertiseEnableChangeAttemptEvent args)
         {
-            // TODO: Improve this...
-            if(args.NewState && component.Broken)
+            if(args.Enabling && component.Broken)
                 args.Cancel();
         }
 
         public override void Update(float frameTime)
         {
-            _timer += frameTime;
-
-            if (_timer < UpdateTimer)
-                return;
-
-            _timer -= UpdateTimer;
-
             var curTime = _gameTiming.CurTime;
+            if (_nextCheckTime > curTime)
+                return;
 
-            foreach (var advertise in EntityManager.EntityQuery<AdvertiseComponent>())
+            _nextCheckTime = curTime + _maximumNextCheckDuration;
+            foreach (var (transform, advertise) in EntityManager.EntityQuery<TransformComponent, AdvertiseComponent>())
             {
                 if (!advertise.Enabled)
                     continue;
 
-                // If it's still not time for the advertisement, do nothing.
                 if (advertise.NextAdvertisementTime > curTime)
+                {
+                    if (advertise.NextAdvertisementTime < _nextCheckTime)
+                    {
+                        _nextCheckTime = advertise.NextAdvertisementTime;
+                    }
                     continue;
+                }
+
+                SayAdvertisement(transform.ParentUid, advertise);
```

Comment:
also i feel like the check for NextAdvertisementTime should be done inside SayAdvertisement instead of here